### ISSUE TITLE
torch.cat should not do type promotion when one input is empty tensor

### ISSUE THREAD

**Issue by tianboh** [pytorch/pytorch#57938]
## 🐛 Bug
This is a edge case for issue #35014

When one of the concatenated tensor is empty, we should not do type promotion. 
## To Reproduce

```
>>> a = torch.zeros(5, dtype=torch.long)
>>> b = torch.tensor([])
>>> c = torch.cat([a,b])
>>> print(a.type())
torch.LongTensor
>>> print(c.type())
torch.FloatTensor
```

## Expected behavior
In fact, empty tensor also has its own type, which is torch.FloatTensor. However, given the 'empty' meaning, we should not change the non-empty tensor type after torch.cat

```
>>> a = torch.zeros(5, dtype=torch.long)
>>> b = torch.tensor([])
>>> c = torch.cat([a,b])
>>> print(a.type())
torch.LongTensor
>>> print(c.type())
torch.LongTensor
```
## Environment
```
PyTorch version: 1.8.1
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 10 Home
GCC version: Could not collect
Clang version: Could not collect
CMake version: Could not collect

Python version: 3.9 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce RTX 2070 SUPER
Nvidia driver version: 461.37
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] botorch==0.4.0
[pip3] gpytorch==1.4.1
[pip3] numpy==1.19.5
[pip3] torch==1.8.1
[pip3] torch-summary==1.4.5
[pip3] torchaudio==0.8.1
[pip3] torchtext==0.9.1
[pip3] torchvision==0.9.1
[conda] blas                      1.0                         mkl    defaults
[conda] botorch                   0.4.0                         0    pytorch
[conda] cudatoolkit               11.1.1               heb2d755_7    conda-forge
[conda] gpytorch                  1.4.1                    pypi_0    pypi
[conda] libblas                   3.9.0                     6_mkl    conda-forge
[conda] libcblas                  3.9.0                     6_mkl    conda-forge
[conda] liblapack                 3.9.0                     6_mkl    conda-forge
[conda] liblapacke                3.9.0                     6_mkl    conda-forge
[conda] mkl                       2020.4             hb70f87d_311    conda-forge
[conda] mkl-service               2.3.0            py39h196d8e1_0    defaults
[conda] mkl_fft                   1.3.0            py39h46781fe_0    defaults
[conda] mkl_random                1.0.2            py39h848d8c7_0    defaults
[conda] numpy                     1.19.5                   pypi_0    pypi
[conda] numpy-base                1.19.2           py39hbd0edd7_0    defaults
[conda] pytorch                   1.8.1           py3.9_cuda11.1_cudnn8_0    pytorch
[conda] torch-summary             1.4.5                    pypi_0    pypi
[conda] torchaudio                0.8.1                      py39    pytorch
[conda] torchtext                 0.9.1                    pypi_0    pypi
[conda] torchvision               0.9.1                py39_cu111    pytorch

```
## Additional context

cc @brianjo @mruberry @nairbv

**Comment by gchanan** [comment]
NumPy type promotes in this case and I don't think it makes sense to complicate the type promotion rules even further:
```
>>> np.concatenate([np.zeros(5, dtype=np.long), np.array([])]).dtype
dtype('float64')
```

**Comment by mruberry** [comment]
I've added the docs label to clarify our documentation to explain this behavior is expected

**Comment by tianboh** [comment]
Thanks! It makes sense now.
